api4/file.go

```diff
@@ -143,6 +149,333 @@ func uploadFile(c *Context, w http.ResponseWriter, r *http.Request) {
+	defer func() {
+		if c.Err != nil {
+			return
+		}
+
+		w.WriteHeader(http.StatusCreated)
+		w.Write([]byte(fileUploadResponse.ToJson()))
+	}()
+
+	if !*c.App.Config().FileSettings.EnableFileAttachments {
+		c.Err = model.NewAppError("uploadFileStream", "api.file.attachments.disabled.app_error",
+			nil, "", http.StatusNotImplemented)
+		return
+	}
+
+	// Parse the post as a regular form (in practice, use the URL values
+	// since we never expect a real application/x-www-form-urlencoded
+	// form).
+	if r.Form == nil {
+		err := r.ParseForm()
+		if err != nil {
+			c.Err = model.NewAppError("uploadFileStream", "api.file.upload_file.read_request.app_error",
+				nil, err.Error(), http.StatusBadRequest)
+			return
+		}
+	}
+
+	timestamp := time.Now()
+
+	_, err := r.MultipartReader()
+	if err == nil {
+		fileUploadResponse = uploadFileMultipart(c, r, timestamp, bufferedMode)
+		return
+	} else if err != http.ErrNotMultipart {
+		c.Err = model.NewAppError("uploadFileStream", "api.file.upload_file.read_request.app_error",
+			nil, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	// Simple POST with the file in the body and all metadata in the args.
+	c.RequireChannelId()
+	c.RequireFilename()
+	if c.Err != nil {
+		return
+	}
+
+	if !c.App.SessionHasPermissionToChannel(c.App.Session, c.Params.ChannelId, model.PERMISSION_UPLOAD_FILE) {
+		c.SetPermissionError(model.PERMISSION_UPLOAD_FILE)
+		return
+	}
+
+	task := c.App.NewUploadFileTask(FILE_TEAM_ID, c.Params.ChannelId, c.App.Session.UserId,
+		c.Params.Filename, timestamp, r.ContentLength, r.Body)
+	task.ClientId = r.Form.Get("client_id")
+	info, appErr := task.Do()
+	if appErr != nil {
+		c.Err = appErr
+		return
+	}
+
+	fileUploadResponse = &model.FileUploadResponse{
+		FileInfos: []*model.FileInfo{info},
+	}
+	if task.ClientId != "" {
+		fileUploadResponse.ClientIds = []string{task.ClientId}
+	}
+}
+
+type uploadFileMultipartMode bool
+
+const (
+	streamMode   = uploadFileMultipartMode(false)
+	bufferedMode = uploadFileMultipartMode(true)
+)
+
+// uploadFileMultipart first peeks into the multipart message to see if
+// channel_id comes before any of the files.  It buffers what it reads out of
+// Input. Then if the channel_id is found, it re-processes the entire POST in
+// the streaming mode, else it prebuffers the entire message, up to the allowed
+// limit.
+func uploadFileMultipart(c *Context, r *http.Request, timestamp time.Time,
+	mode uploadFileMultipartMode) *model.FileUploadResponse {
+
+	expectClientIds := true
+	var clientIds []string
+	resp := model.FileUploadResponse{
+		FileInfos: []*model.FileInfo{},
+		ClientIds: []string{},
+	}
+
+	var buf *bytes.Buffer
+	prevBody := r.Body
+	if mode == bufferedMode {
+		buf = &bytes.Buffer{}
+		// We need to buffer until we get the channel_id, or the first
+		// file.
+		r.Body = ioutil.NopCloser(io.TeeReader(r.Body, buf))
+	}
+
+	// Zero out previous multipartReader so that it can be obtained again.
+	r.MultipartForm = nil
```

Comment:
Right. This is rather tricky. Like you noted before, this complexity is mostly so that we can handle the fully-buffered legacy case, and if/when that code is removed the need for recursion, pre-buffering and all that goes away.

The general logic is,
1. see if the original request is NOT a multipart. process "as is", no buffering.
2. determine if (a) `channel_id` or a (b) file part come first - this necessitates pre-buffering up to that point.
  (2a) reprocess the entire request (pre-buffered + remaining) as multipart, stream
  (2b) reprocess the entire request (pre-buffered + remaining) as multipart, fully buffer

1, 2a, and 2b each use a different `MultipartReader`, wrapping a different Reader each (raw, pre-buffering, and combined). I was tempted to bypass r.MultipartReader altogether for clarity and call `multipart.NewReader` myself - decided against it as it'd require duplicating `Request.multipartReader()`

I agree that it is confusing as is.

I could 
1. Comment it better until the complexity/legacy support are removed
2. Clean up "as is": Move the hacking of `r` and `mr` into a utility function, maybe combine #1 and #2a since there's no overhead in using the pre-buffering reader there
3. Duplicate `Request.multipartReader()` and explicitly obtain one for the right Reader

Neither is a lot of work, I am inclined to try option 3 and see if it ends up less confusing. Let me know if you have a preference. It also feels important to restore the Request to its original `Body` before returning which adds to the complexities here.

Reply:
I refactored as (3) and it is a lot cleaner, please check @lieut-data